File: 2020/julia/AOC/src/day13.jl
module Day13
import AOC.Parsers 

function part1(;data=Parsers.day13())
	start = data[1]
	buses = parse.(Int, filter(x -> x != "x", split(data[2], ',')))
	prod(minimum(map(id -> (id - (start % id), id), buses)))
end

function part2(;data=Parsers.day13())
	splat = split(data[2], ',')
	indexes = findall(x -> x != "x", splat)
	ns = parse.(Int, splat[indexes])
	# Chinese remainder theorem
	N = prod(ns)
	as = [n - (ind-1) for (n, ind) in zip(ns, indexes)]
	ys = [N ÷ n for n in ns]
	zs = [invmod(y, n) for (y, n) in zip(ys, ns)]
	sum([a * y * z for (a, y, z) in zip(as, ys, zs)]) % N
end

end

File: 2020/fsharp/AOC/Day13.fs
module AOC.Day13


let part1 (start, (schedule: string)) = 
    let (diff, bus) = schedule.Split([|','|])
                        |> Seq.filter ((<>) "x")
                        |> Seq.map (int >> (fun bus -> (bus- (start % bus), bus)))
                        |> Seq.minBy (fun (diff, _) -> diff)
    diff * bus

let sieve pairs = 
    let rec catchUp res step off div = 
        if ((res + off) % div) = 0L then res
        else catchUp (res + step) step off div
    let (foff, fdiv) = List.head pairs
    pairs 
    |> Seq.fold (fun (res, step) (off, div) -> ((catchUp res step off div), step * div)) (fdiv-foff, 1L)
    |> fst

let part2 (_, schedule: string) = 
    let pairs = schedule.Split([|','|])
                |> Seq.mapi (fun idx ele -> (idx, ele))
                |> Seq.choose (fun (idx, ele) -> if ele = "x" then None 
                                                    else Some (int64 idx, int64 ele))
                |> List.ofSeq
                |> List.sortByDescending snd
    sieve pairs
                                                    
